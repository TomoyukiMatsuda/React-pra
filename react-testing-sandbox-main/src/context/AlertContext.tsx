import { createContext, useCallback, useContext, useReducer } from "react"

type AlertState = {
  show: boolean
  message: string
}
type AlertDispatch = {
  showDispatcher: (message: string) => void
  hideDispatcher: () => void
}
const initialState = {
  show: false,
  message: ""
}
const alertStateContext = createContext<AlertState>(initialState)
const alertDispatchContext = createContext<AlertDispatch>({
  showDispatcher: () => void 0,
  hideDispatcher: () => void 0
})

export const useAlertState = () => useContext(alertStateContext)
export const useAlertDispatch = () => useContext(alertDispatchContext)

export function AlertProvider({ children }: { children: JSX.Element[] | JSX.Element }) {
  const [state, dispatch] = useReducer(reducer, initialState)
  const showDispatcher = useCallback((message: string) => dispatch(showAlert(message)), [])
  const hideDispatcher = useCallback(() => dispatch(hideAlert()), [])
  return (
    <alertStateContext.Provider value={state}>
      <alertDispatchContext.Provider value={{ showDispatcher, hideDispatcher }}>
        {children}
      </alertDispatchContext.Provider>
    </alertStateContext.Provider>
  )
}

type AlertAction = ReturnType<typeof showAlert> | ReturnType<typeof hideAlert>

function reducer(state: AlertState, action: AlertAction) {
  switch (action.type) {
    case "ui/alert:show":
      return {
        ...state,
        show: true,
        message: action.payload.message
      }
    case "ui/alert:hide":
      return {
        ...state,
        show: false
      }
    default:
      return state
  }
}

function showAlert(message: string) {
  return {
    type: "ui/alert:show",
    payload: {
      message
    }
  } as const
}

function hideAlert() {
  return {
    type: "ui/alert:hide"
  } as const
}
