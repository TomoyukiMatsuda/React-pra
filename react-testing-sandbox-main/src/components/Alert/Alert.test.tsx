import { cleanup, render } from "@testing-library/react"

import * as AlertContext from "~/context/AlertContext"

import { Alert } from "./"

let alertStateSpy: jest.SpyInstance<unknown>

describe("Alert", () => {
  beforeEach(() => {
    alertStateSpy = jest.spyOn(AlertContext, "useAlertState")
    alertStateSpy.mockImplementation(() => ({
      show: false,
      message: ""
    }))
  })
  afterEach(() => {
    alertStateSpy.mockClear()
    cleanup()
  })
  test("render: null", () => {
    const { container } = render(<Alert />)
    expect(container.innerHTML).toBe("")
  })
  test("render: context update", () => {
    const { asFragment, rerender } = render(<Alert />)
    alertStateSpy.mockImplementation(() => ({
      show: true,
      message: "test message"
    }))
    rerender(<Alert />)
    expect(asFragment()).toMatchSnapshot()
  })
})
