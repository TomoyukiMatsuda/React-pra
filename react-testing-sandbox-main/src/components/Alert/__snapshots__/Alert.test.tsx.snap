// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert render: context update 1`] = `
<DocumentFragment>
  <div
    style="position: fixed; top: 0px; left: 0px; width: 100%; padding: 10px; opacity: 0.6; background: red; color: white; font-weight: bold;"
  >
    test message
  </div>
</DocumentFragment>
`;
